ISPN-7578 EntryWrappingInterceptor should not invoke tx commands synchronously

rvansa opened the pull request:
https://issues.jboss.org/browse/ISPN-7578

danberindei commented on core/src/test/java/org/infinispan/tx/EntryWrappingInterceptorDoesNotBlockTest.java:
```diff
@@ -0,0 +1,310 @@
+   }
+
+   public void test1() throws Exception {
+      testMovingStable(this::readWriteKey);
```
How about using a DataProvider, something like this:
https://paste.fedoraproject.org/paste/GwE3F5Cv3sQ9X-V1qwwZp15M1UNdIGYhyRLivL9gydE=

rvansa commented on core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:
```diff
@@ -818,19 +828,37 @@ protected boolean shouldCommitDuringPrepare(PrepareCommand command, TxInvocation
+         c.setTopologyId(command.getTopologyId());
+         InvocationStage visitorStage = makeStage(c.acceptVisitor(ctx, entryWrappingVisitor));
+
+         if (visitorStage.isDone()) {
```
Okay, I don't have a problem with that. I could also add

```
boolean isResultDone(Object result) {
  return  result instanceof InvocationStage ? ((InvocationStage) result).isDone() : true;
}
```
Anyway, I'll change that to variant without instance creation.

danberindei replied:
Oops, I missed both my own comments and your replies when looking at this again.

rvansa commented on core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:
```diff
@@ -818,19 +828,37 @@ protected boolean shouldCommitDuringPrepare(PrepareCommand command, TxInvocation
+               return visitorStage;
+            }
+            return makeStage(visitorStage.thenApply(ctx, command,
+                  (rCtx, rCommand, rv) -> applyModifications(ctx, command, modifications, nextIndex)));
```
Uh, not sure if I follow - you want to refactor that  to a more general solution? Let's leave that for a follow up PR then.

danberindei replied:
Not necessarily a more general solution, I just like making the state explicit so that we know what's being allocated.

rvansa commented on core/src/main/java/org/infinispan/interceptors/totalorder/TotalOrderVersionedEntryWrappingInterceptor.java:
```diff
@@ -32,6 +34,26 @@
    private static final EntryVersionsMap EMPTY_VERSION_MAP = new EntryVersionsMap();
 
+   private final InvocationSuccessFunction prepareHandler
+         = (ctx, command, nil) -> invokeNextThenApply(ctx, command, (rCtx, rCommand, rv) -> {
```
According to what checkstyle?

rvansa commented:
@danberindei updated

rvansa commented:
Rebased & fixed failing tests.

danberindei commented on core/src/test/java/org/infinispan/tx/EntryWrappingInterceptorDoesNotBlockTest.java:
```diff
@@ -0,0 +1,280 @@
+   }
+
+   private Object putDeltaWrite(MagicKey key, int index) {
+      // adding DELTA_WRITE is sufficient to force LoadType.OWNERS, we don't have to do actual delta
```
I guess there's no class check because the parameter to a `DELTA_WRITE` put is a `DeltaAware` on the originator and a `Delta` on the remote nodes. Still, it looks weird :)

danberindei commented:
Got a failure in the new test:

http://ci.infinispan.org/viewLog.html?buildId=55409&buildTypeId=bt9

rvansa commented on core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:
```diff
@@ -817,19 +823,44 @@ protected boolean shouldCommitDuringPrepare(PrepareCommand command, TxInvocation
+
+         InvocationStage visitorStage = null;
+         boolean done = true;
+         if (result instanceof InvocationStage) {
```
Such method does not really fit here; I would have to do the cast in order to throw the exception in the else branch. I could use this && helper method to extract && throw, but I don't find that useful.

danberindei replied:
You don't need to extract and throw explicitly: you just need to know that the stage is not completed successfully and you need to call `makeStage(result).thenApply(...)`, which will throw instead.

rvansa replied:
OK, let's give it a go.

rvansa commented:
@danberindei Rebased, addressed. I've added `asyncInvokeNext(..., Collection<CF<?>> delays)` overload to `BaseAsyncInterceptor` rather than using weird method names.

Wrt the failing test, I was running that locally in a loop and didn't see any failure even without traces; I'll see if CI could reproduce that, and I'll address it with priority if it shows up to be flaky.

danberindei commented:
It's failed every time so far when I ran the core test suite locally, with trace logging enabled.

https://www.dropbox.com/s/1i6k1s3qvjwevne/EntryWrappingInterceptorDoesNotBlockTest_pr_rvansa_ISPN-7578_20170427.log.gz?dl=0

rvansa commented:
Fixed - thanks a lot for the log, @danberindei

galderz commented:
Needs rebasing

rvansa commented:
Rebased.

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/VersionedEntryWrappingInterceptor.java:
```diff
@@ -73,11 +64,27 @@ public Object visitPrepareCommand(TxInvocationContext ctx, PrepareCommand comman
-         return rv;
+         return newVersionData;
       });
+   };
```
Extra comma

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/VersionedEntryWrappingInterceptor.java:
```diff
@@ -29,33 +32,21 @@
-         versionedPrepareCommand.setVersionsSeen(ctx.getCacheTransaction().getVersionsRead());
-      }
-      wrapEntriesForPrepare(ctx, command);
+   private Object prepareHandler(InvocationContext nonTxCtx, VisitableCommand command, Object nil) {
```
I guess forgot to mention why I like having a method for every callback: because it allows us to keep related code together, and to minimize the diff when we extract part of a method into a callback :)

danberindei commented on core/src/main/java/org/infinispan/interceptors/totalorder/TotalOrderVersionedEntryWrappingInterceptor.java:
```diff
@@ -32,6 +35,27 @@
    private static final boolean trace = log.isTraceEnabled();
    private static final EntryVersionsMap EMPTY_VERSION_MAP = new EntryVersionsMap();
 
+   private final InvocationSuccessFunction afterPrepareHandler = (ctx, command, rv) -> {
```
Same here, please make the callback a method after `visitPrepareCommand`.

danberindei commented on core/src/test/java/org/infinispan/tx/EntryWrappingInterceptorDoesNotBlockTest.java:
```diff
@@ -0,0 +1,281 @@
+   }
+
+   @Test(dataProvider = "operations")
+   private void testMovingMoving(Operation operation) throws Exception {
```
Test methods should be public.

rvansa commented:
@danberindei happy?

Dunno why checkstyle does not catch the extra semicolon...

tristantarrant commented:
Great stuff @rvansa , merged to master
